cleverhans/attacks.py

```diff
@@ -1821,3 +1821,109 @@ def loss_fn(x, label):
+            'angle_max': self.np_dtype,
+            'n_angles': self.np_dtype
+        }
+
+    def generate(self, x, **kwargs):
+        """
+        Generate symbolic graph for adversarial examples and return.
+        :param x: The model's symbolic inputs.
+        :param batch_size: (optional int) The size of batch during evaluation.
+        :param n_samples: (optional) The number of transformations sampled to
+                          construct the attack.
+        :param dx_min: (optional float) Minimum translation ratio along x-axis.
+        :param dx_max: (optional float) Maximum translation ratio along x-axis.
+        :param n_dxs: (optional int) Number of discretized translation ratios
+                      along x-axis.
+        :param dy_min: (optional float) Minimum translation ratio along y-axis.
+        :param dy_max: (optional float) Maximum translation ratio along y-axis.
+        :param n_dys: (optional int) Number of discretized translation ratios
+                      along y-axis.
+        :param angle_min: (optional float) Largest counter-clockwise rotation
+                          angle.
+        :param angle_max: (optional float) Largest clockwise rotation angle.
+        :param n_angles: (optional int) Number of discretized angles.
+        """
+        # Parse and save attack-specific parameters
+        assert self.parse_params(**kwargs)
+
+        from .attacks_tf import spm
+
+        labels, _ = self.get_or_guess_labels(x, kwargs)
+
+        return spm(
+            x,
+            self.model,
+            batch_size=self.batch_size,
+            y=labels,
+            n_samples=self.n_samples,
+            dx_min=self.dx_min, dx_max=self.dx_max, n_dxs=self.n_dxs,
+            dy_min=self.dy_min, dy_max=self.dy_max, n_dys=self.n_dys,
+            angle_min=self.angle_min, angle_max=self.angle_max,
+            n_angles=self.n_angles)
+
+    def parse_params(self,
+                     batch_size=128,
+                     n_samples=10,
+                     dx_min=-0.1,
+                     dx_max=0.1,
+                     n_dxs=2,
+                     dy_min=-0.1,
+                     dy_max=0.1,
+                     n_dys=2,
+                     angle_min=-30,
+                     angle_max=30,
+                     n_angles=6,
+                     **kwargs):
+        """
+        Take in a dictionary of parameters and applies attack-specific checks
+        before saving them as attributes.
+        """
+        self.batch_size = batch_size
+        self.n_samples = n_samples
+        self.dx_min = dx_min
+        self.dx_max = dx_max
+        self.n_dxs = n_dxs
+        self.dy_min = dy_min
+        self.dy_max = dy_max
+        self.n_dys = n_dys
+        self.angle_min = angle_min
+        self.angle_max = angle_max
+        self.n_angles = n_angles
+
+        if self.angle_min < -30 or self.angle_max > 30:
+            raise ValueError("The value of angle must be bounded "
```

Comment:
Why do we need to enforce this bound?

Reply:
We don't need to have this, I added it just because this is the bound in the paper.